packages/wrangler/src/dev.tsx

```diff
@@ -565,43 +565,51 @@ function useProxy({
 }) {
   useEffect(() => {
     if (!token) return;
-    const proxy = httpProxy.createProxyServer({
-      secure: false,
-      changeOrigin: true,
-      headers: {
-        "cf-workers-preview-token": token.value,
+    // TODO(soon): since headers are added in callbacks, the server
+    // does not need to restart when changes are made.
+    const host = token.host;
+    const proxy = createHttpProxy({
+      host,
+      assetPath: typeof publicRoot === "string" ? publicRoot : null,
+      onRequest: (headers) => {
+        headers["cf-workers-preview-token"] = token.value;
+      },
+      onResponse: (headers) => {
+        for (const [name, value] of Object.entries(headers)) {
+          // Rewrite the remote host to the local host.
+          if (typeof value === "string" && value.includes(host)) {
+            headers[name] = value
+              .replaceAll(`https://${host}`, `http://localhost:${port}`)
+              .replaceAll(host, `localhost:${port}`);
```

Comment:
Is it OK to mutate the headers object like this?
Would be safer to design this so that the `onResponse(headers)` returns a new headers object with the appropriate changes?

Reply:
It should be okay. But both `onResponse` and `onRequest` could be redesigned to require a return, instead of mutating.

Reply:
happy to revisit this in a future PR, might be a little more involved than just the return from the handler